### ISSUE TITLE
Python 3 functions and lambdas cannot take explicit tuple param

### ISSUE THREAD

**Issue by cclauss** [getsentry/sentry#6030]
The following lines in Sentry can not be ported to Python 3 in their current form...

flake8 testing of https://github.com/getsentry/sentry on Python 3.6.3

$ __flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__
```
./src/sentry/api/endpoints/group_similar_issues.py:21:20: E999 SyntaxError: invalid syntax
            lambda (group_id, scores): group_id != group.id,
                   ^

./src/sentry/api/serializers/models/group.py:78:36: E999 SyntaxError: invalid syntax
                            lambda (project, groups): groups if not options.get(
                                   ^

./src/sentry/digests/backends/redis.py:231:24: E999 SyntaxError: invalid syntax
                lambda (key, value, timestamp): Record(
                       ^

./src/sentry/rules/conditions/event_frequency.py:32:65: E999 SyntaxError: invalid syntax
                      ) in sorted(intervals.items(), key=lambda (key, (label, duration)): duration)
                                                                ^

./src/sentry/similarity/features.py:19:16: E999 SyntaxError: invalid syntax
        lambda (in_app, frames): list(frames),
               ^

./src/sentry/tasks/reports.py:700:26: E999 SyntaxError: invalid syntax
        lambda interval, (project, report): report is not None,
                         ^

./src/sentry/utils/iterators.py:21:20: E999 SyntaxError: invalid syntax
            lambda (i, iterator): advance(i, iterator),
                   ^

./tests/sentry/similarity/test_signatures.py:25:24: E999 SyntaxError: invalid syntax
                lambda (l, r): l == r,
                       ^

8     E999 SyntaxError: invalid syntax
```

**Comment by mattrobenolt** [comment]
Really? Ugh. Does Python 3 have an alternative syntax to achieve the same results?

cc @tkaemming

**Comment by cclauss** [comment]
sentry/src/sentry/tasks/reports.py will require someone better than me!

**Comment by tkaemming** [comment]
> Really? Ugh. Does Python 3 have an alternative syntax to achieve the same results?

Nope. :(

**Comment by dcramer** [comment]
@mattrobenolt you simply would use 'args' or something and slice it

tl;dr writing 'functional' python is going to make it harder and harder to use python 3

**Comment by cclauss** [comment]
Anyone up for taking on:
* sentry/src/sentry/similarity/features.py
* sentry/src/sentry/tasks/reports.py

They were beyond me on the first pass.

**Comment by tkaemming** [comment]
@cclauss Yeah, I'm familiar with those and should be able to clean them up quickly.

**Comment by mattrobenolt** [comment]
It seems we've gone a little excessive on rewriting lambdas. I think the confusion is just the extra parens we tend to use.

For example:

```
foo = lambda (a, b): a + b
```

is a syntax error which you're trying to fix, but:

```
foo = lambda a, b: a + b
```

Works just fine without the parens. So I think most of these changes just require removing the parens. The only thing that doesn't work is argument unpacking. Something like:

```
foo = lambda a, (b, c): a + b + c
```

This syntax is forbidden in python3.

**Comment by cclauss** [comment]
Thanks @mattrobenolt   Great to know about the less aggressive approach.  I believe that I have applied it in all the possible places.

**Comment by cclauss** [comment]
@tkaemming I got the last two done (https://github.com/getsentry/sentry/issues/6030#issuecomment-327247531)  @mattrobenolt Thanks for the "[less aggressive approach](https://github.com/getsentry/sentry/issues/6030#issuecomment-327249654)" -- It simplified things.

**Comment by capissimo** [comment]
Having tested functional capabs, i confronted the same problem. My wayaround looks like this (it really annoys to have to invent gotchas - employing the unnecessary code to convert to iterable & then using next()):

def median(xs):
    return pipe(xs,
                lambda xs: len(xs),
                lambda n: iter((n, n // 2)),
                lambda tup: sorted(xs)[next(tup)] 
                            if next(tup) % 2 == 1 
                            else mean( sorted(xs)[next(tup)-1:][:2] )
               )
